pkg/ds/farm.go

```diff
@@ -0,0 +1,342 @@
+		children:   make(map[fields.ID]*childDS),
+	}
+}
+
+func (dsf *Farm) Start(quitCh <-chan struct{}) {
+	dsf.cleanupDaemonSetPods(quitCh)
+	dsf.mainLoop(quitCh)
+}
+
+const cleanupInterval = 60 * time.Second
+
+// This function removes all pods with a DSIDLabel where the daemon set id does
+// not exist in the store at every interval specified because it is possible
+// that the farm will unexpectedly crash or someone deletes or modifies a node
+func (dsf *Farm) cleanupDaemonSetPods(quitCh <-chan struct{}) {
+	go func() {
+		timer := time.NewTimer(time.Duration(0))
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			timer.Reset(cleanupInterval)
+
+			allDaemonSets, err := dsf.dsStore.List()
+			dsIDMap := make(map[string]ds_fields.DaemonSet)
+			for _, dsFields := range allDaemonSets {
+				dsIDMap[dsFields.ID.String()] = dsFields
+			}
+
+			allPods, err := dsf.applicator.GetMatches(klabels.Everything(), labels.POD)
+			if err != nil {
+				dsf.logger.NoFields().Fatal(err)
+				return
+			}
+
+			for _, podLabels := range allPods {
+				// Only check if it is a pod scheduled by a daemon set
+				if podLabels.Labels.Has(DSIDLabel) {
+					dsID := podLabels.Labels.Get(DSIDLabel)
+
+					// Check if the daemon set exists, if it doesn't unschedule the pod
+					if _, ok := dsIDMap[dsID]; !ok {
+						nodeName, podID, err := labels.NodeAndPodIDFromPodLabel(podLabels)
+						if err != nil {
+							dsf.logger.NoFields().Fatal(err)
+							return
+						}
+
+						// TODO: Since this mirrors the unschedule function in daemon_set.go,
+						// We should find a nice way to couple them together
+						dsf.logger.NoFields().Infof("Unscheduling from %s with dangling daemon set uuid %s", nodeName, dsID)
+
+						_, err = dsf.kpStore.DeletePod(kp.INTENT_TREE, nodeName, podID)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Unable to delete pod from intent tree: %v", err)
+						}
+
+						id := labels.MakePodLabelKey(nodeName, podID)
+						err = dsf.applicator.RemoveLabel(labels.POD, id, DSIDLabel)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Error removing label: %v", err)
+						}
+					}
+				}
+			}
+		}
+	}()
+}
+
+func (dsf *Farm) mainLoop(quitCh <-chan struct{}) {
+	subQuit := make(chan struct{})
+	defer close(subQuit)
+	dsWatch := dsf.dsStore.Watch(subQuit)
+
+	var changes *dsstore.WatchedDaemonSets
+	for {
+		select {
+		case <-quitCh:
+			return
+		case val, ok := <-dsWatch:
+			if !ok {
+				return
+			}
+			changes = val
+			if changes == nil {
+				dsf.logger.Error("Unexpected nil value received on channel")
+				return
+			}
+		}
+
+		for _, child := range dsf.children {
+			select {
+			case <-quitCh:
+				return
+			case err, ok := <-child.errCh:
+				if !ok {
+					// child error channel closed
+					return
+				}
+				dsf.logger.Fatalf("An fatal error has occured in ds: '%v':, %v", child.ds, err)
+			default:
+			}
+		}
+
+		err := dsf.handleDSChanges(quitCh, changes)
```

Comment:
Should we copy/pass the value of the `changes` pointer here instead? This code will pass a copy of the pointer. We could avoid the copy on line 143 in that case.